src/lua/uri.lua

```diff
@@ -62,12 +68,11 @@ local uri_set_stash = buffer.ffi_stash_new('struct uri_set')
 local uri_set_stash_take = uri_set_stash.take
 local uri_set_stash_put = uri_set_stash.put
 
-local function escape_pattern(text)
-    return text:gsub("([^%w])", "%%%1")
-end
-
 local unreserved = function(str)
-    return "^[" .. escape_pattern(str) .. "]$"
+    assert(type(str) == "string")
+    local c_unreserved = ffi.new("char[?]", #str + 1)
+    ffi.copy(c_unreserved, str, #str)
+    return c_unreserved
 end
```

Comment:
Taking this implementation per se (without taking into account, whether given approach is good or bad), I see several strange things:

* Why we should copy it? The FFI passes a Lua string as `const char *` without any problems (https://luajit.org/ext_ffi_semantics.html#convert_fromlua).
* Why we don't copy the leading zero byte?